### PR TITLE
Signature verification

### LINKED ISSUE #175: Signature verification.

### PR DESCRIPTION
**Pull request by kubaplas**
Without ECRecover
Closes #175

### CODE REVIEW

File: mock-vm/types/src/address.rs [hunk 1]
```diff
@@ -11,7 +11,7 @@ pub const CONTRACT_ADDRESS_PREFIX: u32 = 0x0000cadd;
 /// Blockchain-agnostic address representation.
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, BorshSerialize, BorshDeserialize)]
 pub struct Address {
-    data: [u8; ADDRESS_LENGTH]
+    pub data: [u8; ADDRESS_LENGTH]
```
Comment (zie1ony) [inline mock-vm/types/src/address.rs]: I'd not expose it. Add `inner_bytes(&self) -> &[u8];`.

File: modules/src/crypto/mod.rs [hunk 2]
```diff
@@ -8,10 +10,40 @@ pub(crate) fn keccak256(input: &str) -> [u8; 32] {
     result.try_into().unwrap()
 }
 
+/// This function is used to recover the public key from the signature.
+/// The signature needs to be in libsecp256k1 format.
+/// hash is the hash of the message that was signed.
+/// recovery_id is the recovery id of the signature (0 to 3).
+#[allow(dead_code)]
+fn ecrecover(hash: &[u8; 32], signature: &[u8; 64], recovery_id: u8) -> Result<PublicKey, Error> {
```
Comment (zie1ony) [inline modules/src/crypto/mod.rs]: Based on what we discussed earlier, you can remove this in favour of signature verification.

File: modules/src/crypto/mod.rs [hunk 3]
```diff
@@ -8,10 +10,40 @@ pub(crate) fn keccak256(input: &str) -> [u8; 32] {
...
+    let message = libsecp256k1::Message::parse_slice(hash)?;
+    let recovery_id = libsecp256k1::RecoveryId::parse(recovery_id)?;
+    let signature = libsecp256k1::Signature::parse_standard_slice(signature)?;
+    let public_key = libsecp256k1::recover(&message, &signature, &recovery_id)?;
+    Ok(public_key)
+}
+
+fn verify_signature(hash: Bytes, signature: Bytes, public_key: odra::types::PublicKey) -> bool {
```
Comment (zie1ony) [inline modules/src/crypto/mod.rs]: This should be part of `contract_env`. This way you'll have implementations for both mock and casper where they should be and this module could be removed.

File: modules/Cargo.toml [hunk 4]
```diff
@@ -13,6 +13,10 @@ categories = ["wasm"]
 odra = { version = "0.4.0", path = "../core", default-features = false }
 hex = "0.4.3"
 sha3 = "0.10.6"
+libsecp256k1 = "0.7.1"
+
+[dev-dependencies]
```
Comment (zie1ony) [inline modules/Cargo.toml]: What's that?

**Review by zie1ony — CHANGES_REQUESTED** [review]

File: modules/src/crypto/mod.rs (hunk 2 above)
Comment (kubaplas) [reply modules/src/crypto/mod.rs]: Removed

File: mock-vm/types/src/address.rs (hunk 1 above)
Comment (kubaplas) [reply mock-vm/types/src/address.rs]: Ok

File: modules/src/crypto/mod.rs (hunk 3 above)
Comment (kubaplas) [reply modules/src/crypto/mod.rs]: This module stays due to keccak that is used in AccessControl module.

File: modules/Cargo.toml (hunk 4 above)
Comment (kubaplas) [reply modules/Cargo.toml]: Removed alongside ecrecover

File: mock-vm/backend/src/test_env.rs [hunk 5]
```diff
@@ -129,7 +130,21 @@ pub fn gas_report() -> Vec<(String, Balance)> {
...
 }
 
+/// Returns the name of the passed event
 fn extract_event_name(mut bytes: &[u8]) -> Result<String, EventError> {
     let name = BorshDeserialize::deserialize(&mut bytes).map_err(|_| EventError::Formatting)?;
     Ok(name)
 }
+
+/// Returns the public key of the account associated with the given address.
+pub fn public_key(address: &Address) -> PublicKey {
```
Comment (zie1ony) [inline mock-vm/backend/src/test_env.rs]: Why don't you defined `PublicKey` as `PublicKey(Address)`?

File: modules/src/crypto/mod.rs [hunk 6]
```diff
@@ -12,6 +12,8 @@ pub(crate) fn keccak256(input: &str) -> [u8; 32] {
 mod test {
```
Comment (zie1ony) [inline modules/src/crypto/mod.rs]: I think this module should be removed.

**Review by zie1ony — CHANGES_REQUESTED** [review]

File: mock-vm/backend/src/test_env.rs (hunk 5 above)
Comment (kubaplas) [reply mock-vm/backend/src/test_env.rs]: Done

File: modules/src/crypto/mod.rs (hunk 6 above)
Comment (kubaplas) [reply modules/src/crypto/mod.rs]: Removed

File: examples/src/signature_verifier.rs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+#[cfg(test)]
+mod test {
+    use crate::signature_verifier::SignatureVerifierDeployer;
+    use odra::test_env;
+    use odra::types::Bytes;
+
+    #[test]
+    fn signature_verification_works() {
+        let message = "Message to be signed";
```
Comment (zie1ony) [inline examples/src/signature_verifier.rs:27]: try b"Message to be signed"

File: examples/src/signature_verifier.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        // 0x02 for SECP256K1
+        let signature_hex = "021e87e186238fa1df9c222b387a79910388c6ef56285924c7e4f6d7e77ed1d6c61815312cf66a5318db204c693b79e020b1d392dafe8c1b3841e1f6b4c41ca0fa";
+        let signature: [u8; 65] = hex::decode(signature_hex).unwrap().try_into().unwrap();
+        let signature_bytes = &Bytes::from(signature.to_vec());
+
+        // Similar to the above, the public key is tagged:
+        let public_key_hex = "02036d9b880e44254afaf34330e57703a63aec53b5918d4470059b67a4a906350105";
+        let public_key_decoded = hex::decode(public_key_hex).unwrap();
+        let (public_key, _) =
+            casper_types::crypto::PublicKey::from_bytes(public_key_decoded.as_slice()).unwrap();
```
Comment (zie1ony) [inline examples/src/signature_verifier.rs]: use `odra::casper::casper_types`

File: odra-casper/backend/Cargo.toml [hunk 9]
```diff
@@ -19,4 +19,6 @@ casper-types = { workspace = true }
 syn = "1.0.96"
 hex = "0.4.3"
 lazy_static = "1.4.0"
-casper-event-standard = { workspace = true }
\ No newline at end of file
+casper-event-standard = { workspace = true }
+sha256 = "1.1.4"
+ed25519 = "1.2.0"
```
Comment (zie1ony) [inline odra-casper/backend/Cargo.toml]: unneded.

File: odra-casper/backend/Cargo.toml [hunk 10]
```diff
@@ -19,4 +19,6 @@ casper-types = { workspace = true }
 syn = "1.0.96"
 hex = "0.4.3"
 lazy_static = "1.4.0"
-casper-event-standard = { workspace = true }
\ No newline at end of file
+casper-event-standard = { workspace = true }
+sha256 = "1.1.4"
```
Comment (zie1ony) [inline odra-casper/backend/Cargo.toml]: remove

File: mock-vm/backend/src/contract_env.rs [hunk 11]
```diff
@@ -116,3 +116,12 @@ pub fn self_balance() -> Balance {
 pub fn native_token_metadata() -> NativeTokenMetadata {
     NativeTokenMetadata::new()
 }
+
+/// Verifies the signature created using the Backend's default signature scheme.
+pub fn verify_signature(message: &Bytes, signature: &Bytes, public_key: &PublicKey) -> bool {
+    let mut message = message.inner_bytes().clone();
+    message.extend_from_slice(public_key.inner_bytes());
+    let mock_signature_hash = sha256::digest(Bytes::from(message.to_vec()).as_slice());
```
Comment (zie1ony) [inline mock-vm/backend/src/contract_env.rs]: remove hashing here

File: mock-vm/backend/Cargo.toml [hunk 12]
```diff
@@ -17,4 +17,5 @@ odra-mock-vm-types = { path = "../types" }
 ref_thread_local = "0.1.1"
 anyhow = "1.0.65"
 hex = "0.4.3"
-colored = "2.0.0"
\ No newline at end of file
+colored = "2.0.0"
+sha256 = "1.1.4"
```
Comment (zie1ony) [inline mock-vm/backend/Cargo.toml]: remove

**Review by zie1ony — APPROVED** [review]
🔑 🔒 🖊️

**Pull request merged**